enhancements/ingress/tunable-router-buffer-sizes.md

```diff
@@ -0,0 +1,165 @@
+}
+```
+
+The `IngressControllerHeaderBuffer` type has 2 `int` fields, one for
+`HeaderBufferSize` and one for `HeaderBufferMaxRewrite`.
+
+```go
+type IngressControllerHeaderBuffer struct {
+	// buffersize describes how much memory should be reserved
+	// (in bytes) for IngressController connection sessions.
+	//
+	// +kubebuilder:validation:required 
+	// +kubebuilder:validation:minimum=<tbd>
+	// +required
+	HeaderBufferSize int `json:"headerBufferSize,omitempty"`
+
+	// buffermaxrewrite describes how much memory should be reserved
+	// (in bytes) from HeaderBufferSize for HTTP header rewriting
+	// and appending.
+	//
+	// +kubebuilder:validation:required
+	// +kubebuilder:validation:minimum=<tbd>
+	// +required
+	HeaderBufferMaxRewrite int `json:"headerBufferMaxRewrite,omitempty"`
+}
+```
+
+Proper environment variables will be added to the Router deployment to allow for the `HeaderBufferSize` and `HeaderBufferMaxRewrite` values to be
+populated in the HAProxy configuration template.
+
+The [Ingress Operator](https://github.com/openshift/cluster-ingress-operator) will need to be modified to handle both the API changes
+and setting the new Router environment variables.
+
+Integration tests will also need to be written to ensure that the new IngressController API fields work as intended (and that the router
+respects buffer limits without degrading).
+
+### User Stories
+
+#### As a Cluster Administrator, I want to edit the HAProxy configuration by setting "tune.maxrewrite" and "tune.bufsize" so that the Router can accommodate requests with large header sizes
+
+The necessary Router and API changes will be made to allow cluster administrators to tweak these settings on a per-IngressController level.
+
+### Implementation Details/Notes/Constraints
+
+Header buffer size values should be validated to make sure they are at least sane. For example, `HeaderBufferSize` values of `0` or `MAXINT` should most likely
+be rejected.
+
+### Risks and Mitigations
+
+Cluster administrators will have the ability to set IngressController buffer sizes that break cluster ingress.
+It is the responsibility of the administrator to modify the HAProxy buffer values responsibly and appropriately.
+Ideally, the buffer size values would only be modified if absolutely needed.
+
+## Design Details
+
+### Open Questions
+
+How often will cluster administrators take advantage of this enhancement?
+Or, in other words, how sufficient are the current HAProxy buffer default values?
+
+### Test Plan
+
+The ingress operator has end-to-end tests; for this enhancement, a test is added that
+(1) creates an IngressController that specifies HeaderBuffer values (preferable with values that exceed
+the Ingress Controller defaults), (2) sends HTTP requests
+with valid and invalid header length sizes, and (3) verifies that the Ingress Controller only
+drops the requests that have header lengths that exceed the specified HeaderBuffer values.
```

Comment:
An idea:  Use the unique-id header API to verify that setting maxrewrite has the desired behavior.

Reply:
definitely! I mentioned this in this section.